### PR TITLE
Prefill period selector comparison values from URL hash

### LINKED ISSUE #20321: Date comparison selector should stay when changing site
## Expected Behavior
When I select a date comparison other than **Previous period**, then change the selected site, I would like the comparison selector stays at the one selected (and displayed).

## Current Behavior / Steps to Reproduce (for Bugs)
1. Go to your favorite Matomo instance: https://demo.matomo.cloud/index.php
2. Select a date comparison other than **Previous period** (eg. another day in the same month)
3. Change the site selector (from **Demo site** to **Example site** for example)
4. OK, the comparison is still taken into account... But when the comparison date needs to be changed, the date comparison selector has been reset to **Previous period**

## Possible Solution
Minor display bug. Just need to select the date comparison type  again...

## Your Environment
Visible in demo site, but was already buggy in 4.3.1

### PR DESCRIPTION
**Pull request by mneudert**
### Description:

The period selector does not keep information about the last selected comparison type (e.g "previous year" or "custom") across page reloads. Adding a new parameter containing the last type selection solves this issue without guessing the type based on the values of `compareDates` or `comparePeriods` (or introducing custom period parsers).

Fixes #20321.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 1]
```diff
@@ -259,14 +262,16 @@ export default defineComponent({
...
-
     this.isComparing = ComparisonsStore.isComparingPeriods();
     watch(() => ComparisonsStore.isComparingPeriods(), (newVal) => {
       this.isComparing = newVal;
     });
 
+    this.updateSelectedValuesFromHash();
+    this.updateComparisonValuesFromHash();
+    watch(() => MatomoUrl.parsed.value, this.updateSelectedValuesFromHash);
+    watch(() => MatomoUrl.parsed.value, this.updateComparisonValuesFromHash);
```
Comment (michalkleiner) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: Can we put both calls into a single handler for the watch?

**Comment by michalkleiner** [comment]
@mneudert please either remove the Needs review label if the PR is still draft or move the PR from draft to 'full', at the moment I'm not sure if it's ready for review or not.

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 2]
```diff
@@ -464,6 +476,44 @@ export default defineComponent({
...
+      const {
+        params: {
+          date: periodDate,
+          period: periodType,
+        },
+      } = ComparisonsStore.getPeriodComparisons().slice(-1).pop() as PeriodComparison;
+
+      try {
+        Periods.parse(periodType, periodDate);
+      } catch (e) {
```
Comment (caddoo) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]:
```suggestion
      } catch {
```

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 3]
```diff
@@ -464,6 +476,44 @@ export default defineComponent({
...
+
+      this.comparePeriodType = comparePeriodType;
+
+      if (this.comparePeriodType === 'custom') {
+        const periodObj = Periods.get(periodType).parse(periodDate) as Range;
+        const [startDate, endDate] = periodObj.getDateRange();
+
+        this.compareStartDate = format(startDate);
+        this.compareEndDate = format(endDate);
+      }
```
Comment (caddoo) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]:
```suggestion
      if (this.comparePeriodType !== 'custom') {
           return;
      }
      const periodObj = Periods.get(periodType).parse(periodDate) as Range;
      const [startDate, endDate] = periodObj.getDateRange();

      this.compareStartDate = format(startDate);
      this.compareEndDate = format(endDate);
```

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 4]
```diff
@@ -464,6 +476,44 @@ export default defineComponent({
 
       this.setPiwikPeriodAndDate(this.selectedPeriod, this.dateValue!);
     },
+    updateComparisonValuesFromHash() {
```
Comment (caddoo) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: The method name is a little confusing for me, it mentions 'FromHash' but I don't see hash anywhere within the function or passed as a parameter.
Could it be made clearer?

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 5]
```diff
@@ -464,6 +476,44 @@ export default defineComponent({
...
+      if (!COMPARE_PERIOD_TYPES.includes(comparePeriodType)) {
+        return;
+      }
+
+      const {
+        params: {
+          date: periodDate,
+          period: periodType,
+        },
+      } = ComparisonsStore.getPeriodComparisons().slice(-1).pop() as PeriodComparison;
```
Comment (caddoo) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: This code seems to be selecting the last `PeriodComparison` in an array, but it's not clear from reading why, would code as documentation make it clearer? Maybe with a variable name that indicated what it is/why?

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue [hunk 6]
```diff
@@ -259,14 +262,18 @@ export default defineComponent({
...
-    watch(() => MatomoUrl.parsed.value, this.updateSelectedValuesFromHash);
-
     this.isComparing = ComparisonsStore.isComparingPeriods();
     watch(() => ComparisonsStore.isComparingPeriods(), (newVal) => {
       this.isComparing = newVal;
     });
 
+    this.updateSelectedValuesFromHash();
+    this.updateComparisonValuesFromHash();
+    watch(() => MatomoUrl.parsed.value, () => {
```
Comment (caddoo) [inline plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: Just a question, does this watch mean we support changing the URL without page reload, if that's the case have you covered this behaviour in your tests?

File: tests/UI/specs/PeriodSelector_spec.js [hunk 7]
```diff
@@ -134,4 +134,33 @@ describe("PeriodSelector", function () {
 
         expect(await page.screenshotSelector(selector + ',#notificationContainer')).to.matchImage('invalid');
     });
+
+    it('should preselect previous "compare to custom range" on load', async function() {
```
Comment (caddoo) [inline tests/UI/specs/PeriodSelector_spec.js]: I like that you have covered this all with UI tests, however the difference between them all is just the `comparePeriodType` and the first one have `compareDates`.

I'm wondering if this slight differences warrant a full UI test, and they aren't something that can be tested with a unit/integration test? I'm not even sure if we support writing those kind of tests for Vue, but just a thought.

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue (hunk 4 above)
Comment (mneudert) [reply plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: I migrated the naming/watch away from the hash to the (now) single function used from the `ComparisonStore`.

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue (hunk 5 above)
Comment (mneudert) [reply plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: The "why" is (was) mostly an "if a user has entered something, only the last value is what he entered".

Most dashboard UI widgets happily accept multiple periods. But as the period selector itself prevents selecting multiple periods (third, fourth, etc. are getting discarded on apply) I switched to a commented `periods[1]`, ignoring user manipulation.

File: plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue (hunk 6 above)
Comment (mneudert) [reply plugins/CoreHome/vue/src/PeriodSelector/PeriodSelector.vue]: With the changed watch it is now perhaps a bit less obvious, but yes, any hash change was/is directly resulting in a change to the selector content. The changed tests perhaps show that a bit better now as the `about:blank` step was removed.

File: tests/UI/specs/PeriodSelector_spec.js (hunk 7 above)
Comment (mneudert) [reply tests/UI/specs/PeriodSelector_spec.js]: I condensed the screenshots down to the custom range one, that one does the most "work" with filling the date inputs. Other values are verified only with a simple input value match to replace the screenshot to at cover the cases without much overhead.

**Review by caddoo — APPROVED** [review]
Thanks for completing the suggestions, nice work 🎉

**Review by michalkleiner — APPROVED** [review]
Haven't run this locally but the code changes look good to me.

**Comment by sgiehl** [comment]
@mneudert feel free to merge once the conflicts were resolved by rebasing and rebuilding the vue files...

**Pull request merged**